helpers/manifest-parser.js

```diff
@@ -0,0 +1,325 @@
+  'portrait-primary',
+  'portrait-secondary',
+  'landscape-primary',
+  'landscape-secondary'
+];
+
+function parseString(raw, trim) {
+  let value;
+  let warning;
+
+  if (typeof raw === 'string') {
+    value = trim ? raw.trim() : raw;
+  } else {
+    if (raw !== undefined) {
+      warning = 'ERROR: expected a string.';
+    }
+    value = undefined;
+  }
+
+  return {
+    raw,
+    value,
+    warning
+  };
+}
+
+function parseURL(raw) {
+  // TODO: resolve url using baseURL
+  // var baseURL = args.baseURL;
+  // new URL(parseString(raw).value, baseURL);
+  return parseString(raw, true);
+}
+
+function parseColor(raw) {
+  let color = parseString(raw);
+
+  if (color.value === undefined) {
+    return color;
+  }
+
+  // TODO(bckenny): validate color, but for reals
+  // possibly pull in devtools/front_end/common/Color.js
+  // If style.color changes when set to the given color, it is valid. Testing
+  // against 'white' and 'black' in case of the given color is one of them.
+  // var dummy = document.createElement('div');
+  // dummy.style.color = 'white';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'white') {
+  //   return color;
+  // }
+  // dummy.style.color = 'black';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'black') {
+  //   return color;
+  // }
+  return color;
+
+  // color.value = undefined;
+  // color.warning = 'ERROR: color parsing failed';
+
+  // return color;
+}
+
+function parseName(jsonInput) {
+  return parseString(jsonInput.name, true);
+}
+
+function parseShortName(jsonInput) {
+  return parseString(jsonInput.short_name, true);
+}
+
+function parseStartUrl(jsonInput) {
+  // TODO: parse url using manifest_url as a base (missing).
+  // start_url must be same-origin as Document of the top-level browsing context.
+  return parseURL(jsonInput.start_url);
+}
+
+function parseDisplay(jsonInput) {
+  let display = parseString(jsonInput.display, true);
+
+  if (display.value && ALLOWED_DISPLAY_VALUES.indexOf(display.value.toLowerCase()) === -1) {
+    display.value = undefined;
+    display.warning = 'ERROR: \'display\' has an invalid value, will be ignored.';
+  }
+
+  return display;
+}
+
+function parseOrientation(jsonInput) {
+  let orientation = parseString(jsonInput.orientation, true);
+
+  if (orientation.value &&
+      ALLOWED_ORIENTATION_VALUES.indexOf(orientation.value.toLowerCase()) === -1) {
+    orientation.value = undefined;
+    orientation.warning = 'ERROR: \'orientation\' has an invalid value, will be ignored.';
+  }
+
+  return orientation;
+}
+
+function parseIcon(raw) {
+  // TODO: pass manifest url as base.
+  let src = parseURL(raw.src);
+  let type = parseString(raw.type, true);
+
+  let density = {
+    raw: raw.density,
+    value: 1,
+    warning: undefined
+  };
+  if (density.raw !== undefined) {
+    density.value = parseFloat(density.raw);
+    if (isNaN(density.value) || !isFinite(density.value) || density.value <= 0) {
```

Comment:
this is actually somewhat complex, +1 to abstract this into a named concept

@return {boolean}
isValidDensity()